File: frontend/src/types/message.ts
export type FileAttachmentInfo = {
  id: string
  original_filename: string
  file_type: string
  token_estimate: number | null
}

export type Message = {
  id: string
  conversation_id: string
  role: 'user' | 'assistant' | 'system'
  content: string
  created_at: string
  parent_id: string | null
  summary: string | null
  context_tokens: number | null
  files: FileAttachmentInfo[]
}

export type ToolStatus = {
  name: string
  args: Record<string, unknown>
}

export type LocalMessage = Message & {
  streaming?: boolean
  toolStatus?: ToolStatus
  images?: string[]
}

export type TreeNode = {
  id: string
  parent_id: string | null
  role: 'user' | 'assistant' | 'system'
  summary: string | null
  children: TreeNode[]
}

File: backend/app/models/message.py
import uuid
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


class Message(Base):
    __tablename__ = "messages"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    conversation_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("conversations.id", ondelete="CASCADE"), nullable=False, index=True
    )
    parent_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("messages.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    role: Mapped[str] = mapped_column(String(20), nullable=False)   # user | assistant | system
    content: Mapped[str] = mapped_column(Text, nullable=False)
    summary: Mapped[str | None] = mapped_column(String(20), nullable=True)
    token_count: Mapped[int | None] = mapped_column(Integer, nullable=True)
    context_tokens: Mapped[int | None] = mapped_column(Integer, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
